src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2384 @@
+    }
+    
+    // partice size
+    start = this->mStartSize + this->mStartSizeVariance * USFloat::Rand(-1.0f, 1.0f);
+    start = MAX(0, start);
+    
+    particle->mParticleSize = start;
+    
+    end = this->mFinishParticleSize + this->mFinishParticleSizeVariance * USFloat::Rand(-1.0f, 1.0f);
+    particle->mDeltaParticleSize = (end - start) / particle->mTimeToLive;
+    
+    
+    // particle rotation
+    start = this->mRotStart + this->mRotStartVariance * USFloat::Rand(-1.0f, 1.0f);
+    end = this->mRotEnd + this->mRotEndVariance * USFloat::Rand(-1.0f, 1.0f);
+    
+    particle->mParticleRotation = start;
+    particle->mDeltaParticleRotation = (end - start) / particle->mTimeToLive;
+    
+    // position
+    
+    // relative particle position
+    USVec3D loc;
+    if (this->mParticlePositionType == PARTICLE_POSITION_RELATIVE) {
+        loc = this->GetLoc();
+        particle->mStartPosition.Init(loc.mX, loc.mY);
+    }
+    // free particle position
+    else if (this->mParticlePositionType == PARTICLE_POSITION_FREE) {
+        
+        loc.Init(0.0f, 0.0f, 0.0f);
+        USAffine3D modelToWorld = this->GetLocalToWorldMtx ();
+        modelToWorld.Transform(loc);
+        particle->mStartPosition.Init(loc.mX, loc.mY);
+    }
+    
+    // set current position
+    particle->mCurrentPosition.Init(particle->mStartPosition);
+    
+    // direction
+    float a = D2R * (this->mAngle + this->mAngleVariance * USFloat::Rand(-1.0f, 1.0f));
+    
+    
+    // gravity mode
+    if ( this->mEmitterType == EMITTER_GRAVITY ) {
+        USVec2D v;
+        v.Init(Cos(a), Sin(a));
+        
+        float s = this->mSpeed + this->mSpeedVariance * USFloat::Rand(-1.0f, 1.0f);
+        v.Scale( s );
+        
+        // direction
+        particle->mDirection.Init(v);
+        
+        // gravity variance
+        particle->mGravity.mX = this->mGravity[0] + this->mGravityVariance[0] * USFloat::Rand(-1.0f, 1.0f);
+        particle->mGravity.mY = this->mGravity[1] + this->mGravityVariance[1] * USFloat::Rand(-1.0f, 1.0f);
+        
+        // radial accel
+        particle->mRadialAcceleration = this->mRadialAcceleration + this->mRadialAccelVariance * USFloat::Rand(-1.0f, 1.0f);
+        
+        // tangential accel
+        particle->mTangentialAcceleration = this->mTangentialAcceleration + this->mTangentialAccelVariance * USFloat::Rand(-1.0f, 1.0f);
+        
+    }
+    // radial mode
+    else {
+        float startRadius = this->mEndRadius + this->mEndRadiusVariance *  USFloat::Rand(-1.0f, 1.0f);
+        float endRadius = this->mStartRadius + this->mStartRadiusVariance *  USFloat::Rand(-1.0f, 1.0f);
+        
+        particle->mRadius = startRadius;
+        
+        particle->mDeltaRadius = (endRadius - startRadius) / particle->mTimeToLive;
+        
+        particle->mAngle = a;
+        particle->mDegreesPerSecond = D2R * (this->mRotatePerSecond + this->mRotatePerSecondVariance * U
```

Comment:
Oh I see. It's switching the index with the last entry in the array so that when it removes the "last" entry in the array, it's actually removing the particle that used to be in the middle. That's a clever hack, but the current vector implementation here doesn't do the same thing. You'd be better off just changing the implementation to a list and calling `erase`, since `std::list` will do the right thing.